darts/utils/data/tabularization.py

```diff
@@ -165,144 +165,1181 @@ def _create_lagged_data(
+    X = []
+    for i, (series_i, lags_i, min_lag_i, max_lag_i) in enumerate(
+        zip(
+            [target_series, past_covariates, future_covariates],
+            [lags, lags_past_covariates, lags_future_covariates],
+            min_lags,
+            max_lags,
         )
-        len_shortest_features = min(
-            [
-                v
-                for v in [
-                    len_target_features,
-                    len_past_cov_features,
-                    len_future_cov_features,
-                ]
-                if v is not None
+    ):
+        series_and_lags_specified = min_lag_i is not None
+        is_target_series = is_training and (i == 0)
+        if is_target_series or series_and_lags_specified:
+            time_index_i = series_i.time_index
+            # If lags are sufficiently large, `series_i` may not contain all
+            # feature times. For example, if `lags_past_covariates = [-50]`,
+            # then we can construct features for time `51` using the value
+            # of `past_covariates` at time `1`, but `past_covariates` may
+            # only go up to time `30`. This does *not* occur when considering
+            # the target series, however, since this series must have values
+            # for all feature times - these values will become labels.
+            time_idx_too_short = (not is_target_series) and (
+                time_index_i[-1] <= start_time
+            )
+            # If `start_time` not included in `time_index_i`, can 'manually' calculate
+            # what its index *would* be if `time_index_i` were extended to include that time:
+            if time_idx_too_short:
+                start_time_idx = (
+                    len(time_index_i)
+                    - 1
+                    + (start_time - time_index_i[-1]) // series_i.freq
+                )
+            # If `start_time` *is* included in `time_index_i`, need to binary search `time_index_i`
+            # for its position:
+            else:
+                start_time_idx = np.searchsorted(time_index_i, start_time)
+        if series_and_lags_specified:
+            # Windows taken between times `t - max_lag_i` and `t - min_lag_i`
+            window_len = max_lag_i - min_lag_i + 1
+            first_window_start_idx = start_time_idx - max_lag_i
+            # Need `+ 1` since end index is exclusive in Python:
+            first_window_end_idx = start_time_idx - min_lag_i + 1
+            # Other windows are formed by sequentially shifting first window forward
+            # by 1 index position each time; to create `(num_samples - 1)` more windows
+            # in addition to the first window, need to take `(num_samples - 1)` values
+            # after `first_window_end_idx`:
+            vals = series_i.all_values(copy=False)[
+                first_window_start_idx : first_window_end_idx + num_samples - 1, :, :
             ]
+            windows = strided_moving_window(
```

Comment:
This looks really good, and I'm also thinking that at some point we could also think of exposing `stride` to users as a parameter to (indirectly) control the training set size (similar to `max_samples_per_ts`).

Reply:
Great idea - I was originally intending on doing something like that, but I erred against it after seeing how big the PR was getting haha. Maybe something to add in a future PR?

Reply:
Yeah, definitely for later :) I opened an issue here for us not to forget: https://github.com/unit8co/darts/issues/1487